README.txt
= gilenson

* http://github.com/julik/gilenson

== DESCRIPTION:

Обработчик типографских символов в HTML согласно общепринятым правилам.
Посвящается П.Г.Гиленсону[http://www.rudtp.ru/lib.php?book=172], благодаря которому русские правила тех.
редактуры еще как минимум 20 лет останутся бессмысленно старомодными.

Разработчики gilenson - Julik, Mash, Yaroslav Markin.

== ФУНКЦИИ

Gilenson расставит в тексте "умные" правильные кавычки (русские - для кириллицы, английские - для латиницы),
заменит "хитрые" пунктуационные символы на entities и отформатирует знаки типа (c), (tm), телефоны и адреса.

Gilenson базируется на коде Typografica[http://pixel-apes.com/typografica] от PixelApes,
который был приведен к положенному в Ruby стандарту. Основные отличия Gilenson от Typografica на PHP:

* работа только и полностью в UTF-8 (включая entities, применимые в XML)
* поддержка "raw"-вывода (символов вместо entities) - текст выводимый Gilenson можно верстать на бумаге

== ИСПОЛЬЗОВАНИЕ:

Программой gilensize поставляющейся в комплекте
  cat файл_с_текстом_в_utf8.txt | gilensize > красивый_документ.txt

Из кода - как фильтр

  formatter = Gilenson.new
  formatter.configure(:dash=>true)
  for string in strings
    puts formatter.process(string)
  end

или через метод ++gilensize++ для любой строковой переменной

  %{ И вот они таки "приехали"}.gilensize => 'И&#160;вот они&#160;таки &#171;приехали&#187;'

Все дополнительные настройки в таком случае передаются форматтеру

  %{ И вот они таки "приехали"}.gilensize(:laquo=>false) => 'И&#160;вот они&#160;таки "приехали"'

Настройки регулируются через методы
  formatter.dashglue = true
или ассоциированным хешем
  formatter.configure!(:dash=>true, :quotes=>false)

Хеш также можно передавать как последний аргумент методам process и to_html,
в таком случае настройки будут применены только при этом вызове

  beautified = formatter.process(my_text, :dash=>true)

В параметры можно подставить также ключ :all чтобы временно включить или выключить все фильтры

  beautified = formatter.process(my_text, :all=>true)

Помимо этого можно пользоваться каждым фильтром по отдельности используя метод +apply+

Можно менять глифы, которые форматтер использует для подстановок. К примеру,
  formatter.glyph[:nbsp] = '&nbsp;'
заставит форматтер расставлять "традиционные" неразрывные пробелы.
Именно это - большая глупость, но другие глифы заменить может быть нужно.

=== Настройки форматтера

* "inches" - преобразовывать дюймы в знак дюйма;
* "laquo" - кавычки-ёлочки
* "quotes" - кавычки-английские лапки
* "dash" -  проставлять короткое тире (150)
* "emdash" - длинное тире двумя минусами (151)
* "initials" - проставлять тонкие шпации в инициалах
* "copypaste" - замена непечатных и "специальных" юникодных символов на entities
* "(c)" - обрабатывать знак копирайта
* "(r)", "(tm)", "(p)", "+-" - спецсимволы, какие - понятно
* "acronyms" - сворачивание пояснений к аббревиатурам (пояснение - в скобках после аббревиатуры без пробела). Пояснение будет "приклеено" к аббревиатуре полукруглой шпацией.
* "degrees" - знак градуса
* "dashglue", "wordglue" - приклеивание предлогов и дефисов
* "spacing" - запятые и пробелы, перестановка
* "phones" - обработка телефонов
* "html" - при false - запрет использования тагов html
* "de_nobr" - при true все <nobr/> заменяются на <span class="nobr"/>
* "raw_output" - (по умолчанию false) - при true вместо entities выводятся UTF-символы.
  Это нужно чтобы обработанный текст верстать на бумаге.
* "skip_attr" - (по умолчанию false) - при true не отрабатывать типографику в атрибутах тегов (title, alt)
* "skip_code" - (по умолчанию true) - при true не отрабатывать типографику внутри <code/>, <tt/>, CDATA

== УСТАНОВКА:

* sudo gem install gilenson

== ИНТЕГРАЦИЯ С ДРУГИМИ ФОРМАТТЕРАМИ

Если вы используете RuTils в одном приложении со стандартными форматтерами (RedCloth, BlueCloth...) воспользуйтесь
дополнительными модулями которые выводят отформатированный Гиленсоном результат работы форматтеров, но при этом следят
чтобы эти форматтеры не портили текст неподобающим образом (не расставляли неправильные кавычки и так далее).

Gilenson определяет следующие дополнительные форматтеры (они работают так же как и их базовые классы)

* Gilenson::RedClothExtra
* Gilenson::BlueClothExtra
* Gilenson::RDiscountExtra
* Gilenson::MarukuExtra

== ЛИЦЕНЗИЯ:

Copyright (c) 2004-2010 Julik Tarkhanov, Danil Ivanov, Yaroslav Markin and contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Данная лицензия разрешает, безвозмездно, лицам, получившим копию данного программного обеспечения и сопутствующей документации (в дальнейшем именуемыми "Программное Обеспечение"), использовать Программное Обеспечение без ограничений, включая неограниченное право на использование, копирование, изменение, добавление, публикацию, распространение, сублицензирование и/или продажу копий Программного Обеспечения, также как и лицам, которым предоставляется данное Программное Обеспечение, при соблюдении следующих условий:

Вышеупомянутый копирайт и данные условия должны быть включены во все копии или значимые части данного Программного Обеспечения.

ДАННОЕ ПРОГРАММНОЕ ОБЕСПЕЧЕНИЕ ПРЕДОСТАВЛЯЕТСЯ «КАК ЕСТЬ», БЕЗ ЛЮБОГО ВИДА ГАРАНТИЙ, ЯВНО ВЫРАЖЕННЫХ ИЛИ ПОДРАЗУМЕВАЕМЫХ, ВКЛЮЧАЯ, НО НЕ ОГРАНИЧИВАЯСЬ ГАРАНТИЯМИ ТОВАРНОЙ ПРИГОДНОСТИ, СООТВЕТСТВИЯ ПО ЕГО КОНКРЕТНОМУ НАЗНАЧЕНИЮ И НЕНАРУШЕНИЯ ПРАВ. НИ В КАКОМ СЛУЧАЕ АВТОРЫ ИЛИ ПРАВООБЛАДАТЕЛИ НЕ НЕСУТ ОТВЕТСТВЕННОСТИ ПО ИСКАМ О ВОЗМЕЩЕНИИ УЩЕРБА, УБЫТКОВ ИЛИ ДРУГИХ ТРЕБОВАНИЙ ПО ДЕЙСТВУЮЩИМ КОНТРАКТАМ, ДЕЛИКТАМ ИЛИ ИНОМУ, ВОЗНИКШИМ ИЗ, ИМЕЮЩИМ ПРИЧИНОЙ ИЛИ СВЯЗАННЫМ С ПРОГРАММНЫМ ОБЕСПЕЧЕНИЕМ ИЛИ ИСПОЛЬЗОВАНИЕМ ПРОГРАММНОГО ОБЕСПЕЧЕНИЯ ИЛИ ИНЫМИ ДЕЙСТВИЯМИ С ПРОГРАММНЫМ ОБЕСПЕЧЕНИЕМ.


History.txt
=== 1.2.2 / 2011-06-14

* Даже чтение $KCODE на 1.9 вызывает ворнинг.

=== 1.2.0 / 2011-06-14

* Некоторые люди - ебанутые самоубийцы и не пользуются KCODE u. Жизнь их научит.

=== 1.2.0 / 2011-06-14

* Для копирайтов/трейдмарков теперь одна настройка
* Починен символ копирайта на 1.9

=== 1.1.2 / 2010-09-16

* Починен require в тестах на 1.9.2. Курилы не отдадим.

=== 1.1.1 / 2010-09-16

* Восстановлена совместимость с Ruby 1.9

=== 1.1.0 / 2010-08-15

* Модули для интеграции с различными форматтерами теперь живут прямо тут а не в rutils.

=== 1.0.5 / 2010-08-15

* Гиленсон теперь сервируется отдельным блюдом. История изменений ниже по тексту взята из rutils.

=== 1.0.1 - 08.02.2009

* Протестировано на JRuby 1.1.6 (julik)
* Основные шестеренки совместимы с Ruby 1.9.1 (julik)

=== 0.2.4 - 10.07.2008

* В Gilenson закрывающая латинская кавычка не принимается за знак дюйма
* В Gilenson корректно обрабатываются неразрывные пробелы если в сегменте присутствуют пунктуация из глифов
* В Gilenson добавлены настройки force_en_quotes и force_ru_quotes (если нужно свести все кавычки к одной форме)
* До свидания, Gilenson::Obsolete (julik)
* До свидания, <nobr>. Указывайте .nobr { text-wrap: nowrap } в своих стилевых таблицах.

=== 0.2.2 - 24.09.2007

* Gilenson - обрабатываем акронимы по принципу Textile (zajats, julik)
* Gilenson - отформатирован под 2 пробела (julik)

=== 0.2.1 - 02.04.2007

* Восстановлен в gemspec gilenson.rb (Сергей Барабаш)

=== 0.1.9 - 17.02.2007

* Gilenson integration - не съедаем пробелы между тегами при прогоне текста через Textile

=== 0.1.8 - 05.02.2007

* Gilenson - расстановка инициалов не должна цеплять аббревиатуры (julik)

=== 0.1.4 - 02.12.2005

* Gilenson - на Oniguruma сбоит skip_code, но это надо побарывать нормальным токенизатором, пока откладывается (julik)
* Gilenson - в тестах произведена замена <cOdE> на <code> и <TT> на <tt> (mash)
* Gilenson - в двух регулярных выражениях (в "def process_quotes") изменен #{@mark_tag} на
  \#{@mark_tag} чтобы предотвратить segfault ruby 1.8.2 (2004-12-25) на [i386-mswin32]  (mash)
* Gilenson - починен дикий баг с маркером в одинарных кавычках - теперь таки последовательность байтов вместо символов (julik)

=== 0.1.3 - 14.11.2005

* Gilenson - замена троеточий на ellipsis (julik)
* Gilenson - теперь пробел, идущий перед одной или двумя буквами на конце слова
  меняется на glyph[:nbsp]. (mash)
* Gilenson - починен баг из-за которого не типографилась закрывающая кавычка после вопросительного
  и восклицательного знака (julik)
* По умолчанию используется новый Gilenson (Gilenson::Formatter). Старый доступен как Gilenson::Obsolete и через
  String#o_gilensize и грузится только в случае необходимости. Если у вас были тесты провеяющие gilensize
  скорее всего они "полетели" (julik)
* Gilenson - дополнительная фича, не связанная с типографикой: замена амперсанда в следующих атрибутах
  тагов: href, src, data. Пример: <a href="?x=1&y=2&#38;z=3"> => <a href="?x=1&#38;y=2&#38;z=3"> (mash)
* Gilenson - изменён метод process_ampersands (стало process_escape_html), меняющий &,<,> на глифы :amp,:lt,:gt
  при запрете HTML (@settings[:html]=>false). По-умолчанию Гиленсон разрешает использование HTML.
  В иерархии метод выставлен первым для корректной обработки случаев вида
  %{Используйте &#38; вместо &amp;}.gilensize(:html=>false) (mash)
* Gilenson - добавлены :skip_attr=>false и :skip_code=>true для управления обработкой типографики атрибутов тегов title, alt и содержимого <code/>, <tt/>, <![CDATA[...]]> (mash)
* Gilenson - добавлен :raw_output=>true для вывода обычных UTF-8 chars вместо entities (julik)
* Смена структуры модулей Gilenson (julik)
    RuTils::Gilenson::Obsolete - старый класс форматтера
    RuTils::Gilenson::Formatter - новый класс форматтера
    При вызове RuTils::Gilenson.new возвращается новый Formatter
* Gilenson - документация (julik)
* Gilenson - основные группы настроек "по-умолчанию" выносятся в константы (julik)
* Gilenson - фильтры выносятся в методы типа process_фильтр и их можно применять по отдельности через Formatter#apply (julik)
* Gilenson - добавлены &para; и &numero; (последний - не стандарт, но нужен для преобразований символа "номер") (mash)
* Gilenson - добавлен хэш @glyph_copy_paste для подмены юникодных типографских символов на entities
  (опционально, через @settings[:copypaste]=>true) (mash)
* Gilenson - добавлена "склейка" инициалов полукруглой шпацией (thin space)
* Gilenson - добавлен Formatter#process чтобы обрабатывать строки поточно, и сделан удобный доступ к настройкам
  через "хвост" метода (julik)
* Во всех внутренних строковых подстановках Gilenson используются указатели на глифы с двоеточием в начале (julik)
* Gilenson - введен DOUBLE PRIME (&#8243;) для знаков дюйма (julik)
    http://en.wikipedia.org/wiki/Prime_%28mark%29
    Согласно изощренным правилам скошенный prime правильней чем "
* Новый Гиленсон lib/gilenson/gilenson.rb (mash)
    RuTils::GilensonNew::Typografica.new('какой-то текст')
    %q{"какой-то текст"}.n_gilensize

test/test_integration_maruku.rb
# -*- encoding: utf-8 -*-
begin
  require 'maruku'

  # Интеграция с Maruku - Markdown
  class MarukuIntegrationTest < Test::Unit::TestCase
    C = Gilenson::MarukuExtra

    def test_integration_maruku

      assert_equal "<p>И вот&#160;&#171;они пошли туда&#187;, и&#160;шли шли&#160;шли</p>",
        C.new('И вот "они пошли туда", и шли шли шли').to_html

    end
  end
rescue LoadError => boom
  STDERR.puts(boom.message)
end

test/test_integration_bluecloth.rb
# -*- encoding: utf-8 -*-
begin
  require 'bluecloth'

  # Интеграция с BlueCloth - Markdown
  # Сам Markdown никакой обработки типографики не производит (это делает RubyPants, но вряд ли его кто-то юзает на практике)
  class BlueclothIntegrationTest < Test::Unit::TestCase
    C = Gilenson::BlueClothExtra

    def test_integration_markdown
      assert_equal "<p>И вот&#160;&#171;они пошли туда&#187;, и&#160;шли шли&#160;шли</p>",
        C.new('И вот "они пошли туда", и шли шли шли').to_html
    end
  end
rescue LoadError => boom
  STDERR.puts(boom.message)
end

lib/extras/redcloth_extra.rb
# -*- encoding: utf-8 -*-
if RedCloth.class == Module
  # RedCloth 4 - частичная поддержка
  class Gilenson::RedClothExtra < RedCloth::TextileDoc
    def to_html(*anything)
      suspended = super(*anything)

      # Return quotes to original state
      [187, 171, 8220, 8221].map do |e|
        suspended.gsub!(/&\##{e};/, '"')
      end

      # Return spaces to original state
      [160].map do |e|
        suspended.gsub!(/&\##{e};/, ' ')
      end

      suspended.gilensize
    end
  end
else
  # RedCloth 3 - RuTils выполняет перегрузку Textile Glyphs в RedCloth, перенося форматирование спецсимволов на Gilenson.
  # Применять как стандартный RedCloth
  #    RuTils::Gilenson::RedClothExtra.new(some_text).to_html
  class Gilenson::RedClothExtra  < RedCloth
    # Этот метод в RedCloth при наличии Гиленсона становится заглушкой
    def htmlesc(text, mode=0)
      text
    end

    # А этот метод обрабатывает Textile Glyphs - ту самую типографицу.
    # Вместо того чтобы влезать в чужие таблицы мы просто заменим Textile Glyphs на Gilenson - и все будут рады.
    def pgl(text) #:nodoc:
      # Подвешиваем пробелы в начале и в конце каждого блока. Тряпка требует чтобы эти пробелы приехали
      # назад нетронутыми.
      spaces =  Array.new(2,'')
      text.gsub!(/\A([\s]+)/) { spaces[0] = $1; '' }
      text.gsub!(/(\s+)\Z/) { spaces[1] = $1; '' }
      text.replace(spaces[0].to_s + ::Gilenson.new(text).to_html + spaces[1].to_s)
    end
  end
end

lib/gilenson/config.rb
class Config
  SETTINGS = {
    "inches"    => true,    # преобразовывать дюймы в знак дюйма;
    "laquo"     => true,    # кавычки-ёлочки
    "quotes"    => true,    # кавычки-английские лапки
    "dash"      => true,    # короткое тире (150)
    "emdash"    => true,    # длинное тире двумя минусами (151)
    "initials"  => true,    # тонкие шпации в инициалах
    "copypaste" => false,   # замена непечатных и "специальных" юникодных символов на entities
    "(c)" => true,    # обрабатывать знаки копирайтов и трейдмарков
    "acronyms"  => true,    # Акронимы с пояснениями - ЖЗЛ(Жизнь Замечатльных Людей)
    "+-"        => true,    # спецсимволы, какие - понятно
    "degrees"   => true,    # знак градуса
    "dashglue"  => true, "wordglue" => true, # приклеивание предлогов и дефисов
    "spacing"   => true,    # запятые и пробелы, перестановка
    "phones"    => true,    # обработка телефонов
    "html"      => true,    # разрешение использования тагов html
    "raw_output" => false,  # выводить UTF-8 вместо entities
    "skip_attr" => false,   # при true не отрабатывать типографику в атрибутах тегов
    "skip_code" => true,    # при true не отрабатывать типографику внутри <code/>, <tt/>, CDATA
    "enforce_en_quotes" => false, # только латинские кавычки
    "enforce_ru_quotes" => false, # только русские кавычки (enforce_en_quotes при этом игнорируется)
    "(r)"       => true, # DEPRECATED
    "(tm)"      => true, # DEPRECATED
    "(p)"       => true, # DEPRECATED
  }.freeze

  #:stopdoc:
  # Старые настройки которые больше не нужны - уйдут в следующей большой версии
  # Нужны потому иначе будет брошена ошибка
  #:startdoc:

  # Глифы, использующиеся в подстановках по-умолчанию
  GLYPHS = {
    :quot       => "&#34;",     # quotation mark
    :amp        => "&#38;",     # ampersand
    :apos       => "&#39;",     # apos
    :gt         => "&#62;",     # greater-than sign
    :lt         => "&#60;",     # less-than sign
    :nbsp       => "&#160;",    # non-breaking space
    :sect       => "&#167;",    # section sign
    :copy       => "&#169;",    # copyright sign
    :laquo      => "&#171;",    # left-pointing double angle quotation mark = left pointing guillemet
    :reg        => "&#174;",    # registered sign = registered trade mark sign
    :deg        => "&#176;",    # degree sign
    :plusmn     => "&#177;",    # plus-minus sign = plus-or-minus sign
    :para       => "&#182;",    # pilcrow sign = paragraph sign
    :middot     => "&#183;",    # middle dot = Georgian comma = Greek middle dot
    :raquo      => "&#187;",    # right-pointing double angle quotation mark = right pointing guillemet
    :ndash      => "&#8211;",   # en dash
    :mdash      => "&#8212;",   # em dash
    :lsquo      => "&#8216;",   # left single quotation mark
    :rsquo      => "&#8217;",   # right single quotation mark
    :ldquo      => "&#8220;",   # left double quotation mark
    :rdquo      => "&#8221;",   # right double quotation mark
    :bdquo      => "&#8222;",   # double low-9 quotation mark
    :bull       => "&#8226;",   # bullet = black small circle
    :hellip     => "&#8230;",   # horizontal ellipsis = three dot leader
    :numero     => "&#8470;",   # numero
    :trade      => "&#8482;",   # trade mark sign
    :minus      => "&#8722;",   # minus sign
    :inch       => "&#8243;",   # inch/second sign (u0x2033) (не путать с кавычками!)
    :thinsp     => "&#8201;",   # полукруглая шпация (тонкий пробел)
    :nob_open   => '<span class="nobr">',    # открывающий блок без переноса слов
    :nob_close  => '</span>',    # закрывающий блок без переноса слов
  }.freeze

  # Нормальные "типографские" символы в UTF-виде. Браузерами обрабатываются плохонько, поэтому
  # лучше заменять их на entities.
  VERBATIM_GLYPHS = {
    ' '         => :nbsp,# alt+0160 (NBSP here)
    '«'         => :laquo,
    '»'         => :raquo,
    '§'         => :sect,
    '©'         => :copy,
    '®'         => :reg,
    '°'         => :deg,
    '±'         => :plusmn,
    '¶'         => :para,
    '·'         => :middot,
    '–'         => :ndash,
    '—'         => :mdash,
    '‘'         => :lsquo,
    '’'         => :rsquo,
    '“'         => :ldquo,
    '”'         => :rdquo,
    '„'         => :bdquo,
    '•'         => :bull,
    '…'         => :hellip,
    '№'         => :numero,
    '™'         => :trade,
    '−'         => :minus,
    ' '         => :thinsp,
    '″'         => :inch,
  }.freeze

  # Метка на которую подменяются вынутые теги  - Unicode Character 'OBJECT REPLACEMENT CHARACTER' (U+FFFC)
  # http://unicode.org/reports/tr20/tr20-1.html
  # Он официально применяется для обозначения вложенного обьекта
  REPLACEMENT_MARKER = [0xEF, 0xBF, 0xBC].pack("U*").freeze

  # Кто придумал &#147;? Не учите людей плохому...
  # Привет А.Лебедеву http://www.artlebedev.ru/kovodstvo/62/
  # Используем символы, потом берем по символам из glyphs форматтера.
  # Молодец mash!
  FORBIDDEN_NUMERIC_ENTITIES = {
    '132'       => :bdquo,
    '133'       => :hellip,
    '146'       => :apos,
    '147'       => :ldquo,
    '148'       => :rdquo,
    '149'       => :bull,
    '150'       => :ndash,
    '151'       => :mdash,
    '153'       => :trade,
  }.freeze

  # Все юникодные пробелы, шпации и отступы
  UNICODE_WHITESPACE = [
    (0x0009..0x000D).to_a, # White_Space # Cc   [5] <control-0009>..<control-000D>
    0x0020,                # White_Space # Zs       SPACE
    0x0085,                # White_Space # Cc       <control-0085>
    0x00A0,                # White_Space # Zs       NO-BREAK SPACE
    0x1680,                # White_Space # Zs       OGHAM SPACE MARK
    0x180E,                # White_Space # Zs       MONGOLIAN VOWEL SEPARATOR
    (0x2000..0x200A).to_a, # White_Space # Zs  [11] EN QUAD..HAIR SPACE
    0x2028,                # White_Space # Zl       LINE SEPARATOR
    0x2029,                # White_Space # Zp       PARAGRAPH SEPARATOR
    0x202F,                # White_Space # Zs       NARROW NO-BREAK SPACE
    0x205F,                # White_Space # Zs       MEDIUM MATHEMATICAL SPACE
    0x3000,                # White_Space # Zs       IDEOGRAPHIC SPACE
  ].flatten.pack("U*").freeze

  PROTECTED_SETTINGS = [ :raw_output ] #:nodoc:

  SKIP_CODE_TAGS = ["code", "tt"]

  def initialize(config = {})
    accept_configuration_arguments!(config)

    @ignore = /notypo/ # regex, который игнорируется. Этим надо воспользоваться для обработки pre и code

    @glueleft =  ['рис.', 'табл.', 'см.', 'им.', 'ул.', 'пер.', 'кв.', 'офис', 'оф.', 'г.']
    @glueright = ['руб.', 'коп.', 'у.е.', 'мин.']

    @mark_tag = REPLACEMENT_MARKER

    @settings = SETTINGS.dup

    # Глифы можено подменять в экземпляре форматтера поэтому копируем их из константы
    @glyph = GLYPHS.dup

    @phonemasks = [
      [
        /([0-9]{4})\-([0-9]{2})\-([0-9]{2}) ([0-9]{2}):([0-9]{2}):([0-9]{2})/,
        /([0-9]{4})\-([0-9]{2})\-([0-9]{2})/,
        /(\([0-9\+\-]+\)) ?([0-9]{3})\-([0-9]{2})\-([0-9]{2})/,
        /(\([0-9\+\-]+\)) ?([0-9]{2})\-([0-9]{2})\-([0-9]{2})/,
        /(\([0-9\+\-]+\)) ?([0-9]{3})\-([0-9]{2})/,
        /(\([0-9\+\-]+\)) ?([0-9]{2})\-([0-9]{3})/,
        /([0-9]{3})\-([0-9]{2})\-([0-9]{2})/,
        /([0-9]{2})\-([0-9]{2})\-([0-9]{2})/,
        /([0-9]{1})\-([0-9]{2})\-([0-9]{2})/,
        /([0-9]{2})\-([0-9]{3})/,
        /([0-9]+)\-([0-9]+)/,
      ],[
        ':nob_open\1:ndash\2:ndash\3:nbsp\4:\5:\6:nob_close',
        ':nob_open\1:ndash\2:ndash\3:nob_close',
        ':nob_open\1:nbsp\2:ndash\3:ndash\4:nob_close',
        ':nob_open\1:nbsp\2:ndash\3:ndash\4:nob_close',
        ':nob_open\1:nbsp\2:ndash\3:nob_close',
        ':nob_open\1:nbsp\2:ndash\3:nob_close',
        ':nob_open\1:ndash\2:ndash\3:nob_close',
        ':nob_open\1:ndash\2:ndash\3:nob_close',
        ':nob_open\1:ndash\2:ndash\3:nob_close',
        ':nob_open\1:ndash\2:nob_close',
        ':nob_open\1:ndash\2:nob_close'
      ]
    ]
  end

  # Настраивает форматтер ассоциированным хешем
  #  formatter.configure!(:dash=>true, :wordglue=>false)
  def configure!(*config)
    accept_configuration_arguments!(config.last) if config.last.is_a?(Hash)
  end

  def accept_configuration_arguments!(args_hash)
    # Специальный случай - :all=>true|false
    if args_hash.has_key?(:all)
      if args_hash[:all]
        @settings.each_pair {|k, v| @settings[k] = true unless PROTECTED_SETTINGS.include?(k.to_sym)}
      else
        @settings.each_pair {|k, v| @settings[k] = false unless PROTECTED_SETTINGS.include?(k.to_sym)}
      end
    else
      # Кинуть ошибку если настройка нам неизвестна
      unknown_settings = args_hash.keys.collect{|k|k.to_s} - @settings.keys.collect { |k| k.to_s }
      raise UnknownSetting, unknown_settings if unknown_settings.any?

      args_hash.each_pair do | key, value |
        @settings[key.to_s] = (value ? true : false)
      end
    end
  end

  alias :configure :configure! #Дружественный API

  # Неизвестные методы - настройки. С = - установка ключа, без - получение значения
  def method_missing(meth, *args) #:nodoc:
    setting = meth.to_s.gsub(/=$/, '')
    super(meth, *args) unless @settings.has_key?(setting) #this will pop the exception if we have no such setting

    return (@settings[setting] = args[0])
  end
end